x-pack/plugins/actions/server/config.ts

```diff
@@ -68,10 +77,32 @@ export const configSchema = schema.object({
   }),
   proxyUrl: schema.maybe(schema.string()),
   proxyHeaders: schema.maybe(schema.recordOf(schema.string(), schema.string())),
+  /**
+   * @deprecated in favor of `ssl.proxyVerificationMode`
+   **/
   proxyRejectUnauthorizedCertificates: schema.boolean({ defaultValue: true }),
   proxyBypassHosts: schema.maybe(schema.arrayOf(schema.string({ hostname: true }))),
   proxyOnlyHosts: schema.maybe(schema.arrayOf(schema.string({ hostname: true }))),
+  /**
+   * @deprecatedin favor of `ssl.verificationMode`
+   **/
   rejectUnauthorized: schema.boolean({ defaultValue: true }),
+  tls: schema.maybe(
+    schema.object({
+      verificationMode: schema.maybe(
```

Comment:
If users can set `rejectUnauthorized` and `verificationMode` at the same time, it may lead to confusion because `verificationMode` take precedence. It may be a better idea to prevent users from setting these both at the same time.

GitHub isn't letting me suggest changes for all of the lines, but I've added a very ugly diff below to explain what I mean!

<details>
<summary>Expand diff</summary>

```diff
diff --git a/x-pack/plugins/actions/server/config.ts b/x-pack/plugins/actions/server/config.ts
index dbd0fd18a55..ff03224b7b1 100644
--- a/x-pack/plugins/actions/server/config.ts
+++ b/x-pack/plugins/actions/server/config.ts
@@ -32,87 +32,108 @@ const customHostSettingsSchema = schema.object({
     })
   ),
   tls: schema.maybe(
-    schema.object({
-      /**
-       * @deprecated in favor of `verificationMode`
-       **/
-      rejectUnauthorized: schema.maybe(schema.boolean()),
-      verificationMode: schema.maybe(
-        schema.oneOf(
-          [schema.literal('none'), schema.literal('certificate'), schema.literal('full')],
-          { defaultValue: 'full' }
-        )
-      ),
-      certificateAuthoritiesFiles: schema.maybe(
-        schema.oneOf([
-          schema.string({ minLength: 1 }),
-          schema.arrayOf(schema.string({ minLength: 1 }), { minSize: 1 }),
-        ])
-      ),
-      certificateAuthoritiesData: schema.maybe(schema.string({ minLength: 1 })),
-    })
+    schema.object(
+      {
+        /**
+         * @deprecated in favor of `verificationMode`
+         **/
+        rejectUnauthorized: schema.maybe(schema.boolean()),
+        verificationMode: schema.maybe(
+          schema.oneOf(
+            [schema.literal('none'), schema.literal('certificate'), schema.literal('full')],
+            { defaultValue: 'full' }
+          )
+        ),
+        certificateAuthoritiesFiles: schema.maybe(
+          schema.oneOf([
+            schema.string({ minLength: 1 }),
+            schema.arrayOf(schema.string({ minLength: 1 }), { minSize: 1 }),
+          ])
+        ),
+        certificateAuthoritiesData: schema.maybe(schema.string({ minLength: 1 })),
+      },
+      {
+        validate: ({ rejectUnauthorized, verificationMode }) => {
+          if (rejectUnauthorized !== undefined && verificationMode) {
+            return 'cannot use [rejectUnauthorized] when [verificationMode] is specified; instead, you should only use [verificationMode]';
+          }
+        },
+      }
+    )
   ),
 });
 
 export type CustomHostSettings = TypeOf<typeof customHostSettingsSchema>;
 
-export const configSchema = schema.object({
-  enabled: schema.boolean({ defaultValue: true }),
-  allowedHosts: schema.arrayOf(
-    schema.oneOf([schema.string({ hostname: true }), schema.literal(AllowedHosts.Any)]),
-    {
-      defaultValue: [AllowedHosts.Any],
-    }
-  ),
-  enabledActionTypes: schema.arrayOf(
-    schema.oneOf([schema.string(), schema.literal(EnabledActionTypes.Any)]),
-    {
-      defaultValue: [AllowedHosts.Any],
-    }
-  ),
-  preconfiguredAlertHistoryEsIndex: schema.boolean({ defaultValue: false }),
-  preconfigured: schema.recordOf(schema.string(), preconfiguredActionSchema, {
-    defaultValue: {},
-    validate: validatePreconfigured,
-  }),
-  proxyUrl: schema.maybe(schema.string()),
-  proxyHeaders: schema.maybe(schema.recordOf(schema.string(), schema.string())),
-  /**
-   * @deprecated in favor of `ssl.proxyVerificationMode`
-   **/
-  proxyRejectUnauthorizedCertificates: schema.boolean({ defaultValue: true }),
-  proxyBypassHosts: schema.maybe(schema.arrayOf(schema.string({ hostname: true }))),
-  proxyOnlyHosts: schema.maybe(schema.arrayOf(schema.string({ hostname: true }))),
-  /**
-   * @deprecatedin favor of `ssl.verificationMode`
-   **/
-  rejectUnauthorized: schema.boolean({ defaultValue: true }),
-  tls: schema.maybe(
-    schema.object({
-      verificationMode: schema.maybe(
-        schema.oneOf(
-          [schema.literal('none'), schema.literal('certificate'), schema.literal('full')],
-          { defaultValue: 'full' }
-        )
-      ),
-      proxyVerificationMode: schema.maybe(
-        schema.oneOf(
-          [schema.literal('none'), schema.literal('certificate'), schema.literal('full')],
-          { defaultValue: 'full' }
-        )
-      ),
-    })
-  ),
-  maxResponseContentLength: schema.byteSize({ defaultValue: '1mb' }),
-  responseTimeout: schema.duration({ defaultValue: '60s' }),
-  customHostSettings: schema.maybe(schema.arrayOf(customHostSettingsSchema)),
-  cleanupFailedExecutionsTask: schema.object({
+export const configSchema = schema.object(
+  {
     enabled: schema.boolean({ defaultValue: true }),
-    cleanupInterval: schema.duration({ defaultValue: '5m' }),
-    idleInterval: schema.duration({ defaultValue: '1h' }),
-    pageSize: schema.number({ defaultValue: 100 }),
-  }),
-});
+    allowedHosts: schema.arrayOf(
+      schema.oneOf([schema.string({ hostname: true }), schema.literal(AllowedHosts.Any)]),
+      {
+        defaultValue: [AllowedHosts.Any],
+      }
+    ),
+    enabledActionTypes: schema.arrayOf(
+      schema.oneOf([schema.string(), schema.literal(EnabledActionTypes.Any)]),
+      {
+        defaultValue: [AllowedHosts.Any],
+      }
+    ),
+    preconfiguredAlertHistoryEsIndex: schema.boolean({ defaultValue: false }),
+    preconfigured: schema.recordOf(schema.string(), preconfiguredActionSchema, {
+      defaultValue: {},
+      validate: validatePreconfigured,
+    }),
+    proxyUrl: schema.maybe(schema.string()),
+    proxyHeaders: schema.maybe(schema.recordOf(schema.string(), schema.string())),
+    /**
+     * @deprecated in favor of `ssl.proxyVerificationMode`
+     **/
+    proxyRejectUnauthorizedCertificates: schema.boolean({ defaultValue: true }),
+    proxyBypassHosts: schema.maybe(schema.arrayOf(schema.string({ hostname: true }))),
+    proxyOnlyHosts: schema.maybe(schema.arrayOf(schema.string({ hostname: true }))),
+    /**
+     * @deprecatedin favor of `ssl.verificationMode`
+     **/
+    rejectUnauthorized: schema.boolean({ defaultValue: true }),
+    tls: schema.maybe(
+      schema.object({
+        verificationMode: schema.maybe(
+          schema.oneOf(
+            [schema.literal('none'), schema.literal('certificate'), schema.literal('full')],
+            { defaultValue: 'full' }
+          )
+        ),
+        proxyVerificationMode: schema.maybe(
+          schema.oneOf(
+            [schema.literal('none'), schema.literal('certificate'), schema.literal('full')],
+            { defaultValue: 'full' }
+          )
+        ),
+      })
+    ),
+    maxResponseContentLength: schema.byteSize({ defaultValue: '1mb' }),
+    responseTimeout: schema.duration({ defaultValue: '60s' }),
+    customHostSettings: schema.maybe(schema.arrayOf(customHostSettingsSchema)),
+    cleanupFailedExecutionsTask: schema.object({
+      enabled: schema.boolean({ defaultValue: true }),
+      cleanupInterval: schema.duration({ defaultValue: '5m' }),
+      idleInterval: schema.duration({ defaultValue: '1h' }),
+      pageSize: schema.number({ defaultValue: 100 }),
+    }),
+  },
+  {
+    validate: ({ rejectUnauthorized, proxyRejectUnauthorizedCertificates, tls }) => {
+      if (rejectUnauthorized !== undefined && tls?.verificationMode) {
+        return 'cannot use [rejectUnauthorized] when [tls.verificationMode] is specified; instead, you should only use [tls.verificationMode]';
+      }
+      if (proxyRejectUnauthorizedCertificates !== undefined && tls?.proxyVerificationMode) {
+        return 'cannot use [proxyRejectUnauthorizedCertificates] when [tls.proxyVerificationMode] is specified; instead, you should only use [tls.proxyVerificationMode]';
+      }
+    },
+  }
+);
 
 export type ActionsConfig = TypeOf<typeof configSchema>;
 

```

</details>

Alternatively, you could log a warning like the one used below: https://github.com/elastic/kibana/blob/a105c7a8b1a028df604fe6d3ccd1af9087cab097/x-pack/plugins/actions/server/config.ts#L95-L98

Reply:
I think, what Larry suggested below, could help to solve this issue.

Reply:
@jportner do you think we still need this validation with the changes I've provided in the [commit](https://github.com/elastic/kibana/pull/100179/commits/de40bf29727020959ead6f100fb0c724f60ef09f)?

Reply:
Yeah, using a deprecation may be a better idea than a regular warning log. But the deprecation warning that's been added mentions that the config option is unused -- that's not quite true, and it also doesn't mention that `verificationMode ` effectively overrides `rejectUnauthorized`.

That's why I didn't suggest adding a deprecation originally as there isn't a 1:1 mapping so you can't use `renameFromRoot()` for it, but `unused()` feels wrong too.

Reply:
I agree, that existing deprecation options `renameFromRoot()` and `unused()` don't fit well enough for the changes I've done. I will replace it with the custom deprecation function with the proper message.

Reply:
@jportner I've added the [commit](https://github.com/elastic/kibana/pull/100179/commits/1155f1598fdf51bd0ce2b1d6ecf7cadee1c94625) where switched from `unsed()` deprecation mode to a custom function with the specific message. Do you think it will enough to explain this to the users?